### PR TITLE
Add mismatches to exprs

### PR DESCRIPTION
**Pull request by dylant-da**

### CODE REVIEW

**Review by samuel-williams-da — COMMENTED** [review]
Forgot to send these from yesterday

File: sdk/compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Alpha.hs [hunk 1]
```diff
@@ -61,260 +154,264 @@ bindExprVar x1 x2 env@AlphaEnv{..} = env
...
+            -> alphaTypeCon env t1 t2
+            && alphaEq' ch1 ch2
             && alphaExpr' env e1a e2a
             && alphaExpr' env e1b e2b
-        _ -> False
+        _ -> structuralMismatch
     EExperimental n1 t1 -> \case
-        EExperimental n2 t2 -> n1 == n2 && alphaType t1 t2
-        _ -> False
+        EExperimental n2 t2 -> n1 `eqStructural` n2 && alphaType' initialAlphaEnv t1 t2 -- TODO: Should alphaType receive env here?
```
Comment (samuel-williams-da) [inline sdk/compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Alpha.hs:413]: Don't forget this TODO :)

File: sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs [hunk 2]
```diff
@@ -785,17 +790,26 @@ warningLocation = \case
...
+    WTemplateChangedKeyMaintainers template mismatches -> withMismatchInfo mismatches $ "The upgraded template " <> pPrint template <> " has changed the maintainers for its key."
     WCouldNotExtractForUpgradeChecking attribute mbExtra -> "Could not check if the upgrade of " <> text attribute <> " is valid because its expression is the not the right shape." <> foldMap (const " Extra context: " <> text) mbExtra
     WErrorToWarning err -> pPrint err
+    where
+    withMismatchInfo :: [Mismatch] -> Doc ann -> Doc ann
+    withMismatchInfo [] doc = doc
+    withMismatchInfo mismatches doc =
+      vcat
+        [ doc
+        , "There are " <> string (show (length mismatches)) <> " differences in the expression, including:"
```
Comment (samuel-williams-da) [inline sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs:810]: "There are 1 differences in the expression" is kinda ugly, might be worth fixing, up to you

File: sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs [hunk 3]
```diff
@@ -27,15 +28,37 @@ import           Data.List (foldl', nub)
...
+      Just udPackageVersion -> " (v" <> show udPackageVersion <> ")"
+      Nothing -> mempty
+
+type UpgradeablePackage = (Maybe RawPackageVersion, LF.PackageId, LF.Package)
+type UpgradeablePackageMap = HMS.HashMap LF.PackageName [UpgradeablePackage]
+
+--data UpgradeablePackageVersion
+--  = NotUpgradeable
+--  | UpgradeablePackageVersion RawPackageVersion
+--  deriving (Show, Eq, Ord)
```
Comment (samuel-williams-da) [inline sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs:61]: Remember to delete

File: sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs [hunk 4]
```diff
@@ -812,79 +833,129 @@ checkType type_ err = do
...
+        Upgrading { _past = Nothing, _present = Nothing } ->
+          [nameMismatch' ("Name came from package '" <> tryShowPkgId (_past pkgId) <> "' and now comes from package '" <> tryShowPkgId (_present pkgId) <> "'. Could not find " <> tryShowPkgId (_past pkgId) <> " nor " <> tryShowPkgId (_present pkgId) <> " in the package list.")]
+        Upgrading { _past = Just _, _present = Nothing } ->
+          [nameMismatch' ("Name came from package '" <> tryShowPkgId (_past pkgId) <> "' and now comes from package '" <> tryShowPkgId (_present pkgId) <> "'. Could not find " <> tryShowPkgId (_present pkgId) <> " in the package list.")]
+        Upgrading { _past = Nothing, _present = Just _ } ->
+          [nameMismatch' ("Name came from package '" <> tryShowPkgId (_past pkgId) <> "' and now comes from package '" <> tryShowPkgId (_present pkgId) <> "'. Could not find " <> tryShowPkgId (_past pkgId) <> " in the package list.")]
+
+checkQualType :: UpgradingDeps -> Upgrading (Qualified TypeConName) -> [Mismatch]
```
Comment (samuel-williams-da) [inline sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs:896]: Given the size of these functions now, it probably makes sense to try dedup. The `checkQualType` differs in that it doesn't check utility packages, but even if it had the code to do so, it'd never hit that branch. Could you not simply use the same (ish) function for both?

**Review by samuel-williams-da — COMMENTED** [review]
The rest of my review :)

File: sdk/compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Alpha.hs (hunk 1 above)
Comment (dylant-da) [reply sdk/compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Alpha.hs:413]: Opted to pass down the environment

File: sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs (hunk 4 above)
Comment (dylant-da) [reply sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs:896]: The main issue is printing different messages for types vs values. For values, the message is a warning, and worded as such. For types, not so much.

**Review by samuel-williams-da — APPROVED** [review]

**Pull request merged**